### PR TITLE
Extend Submission Requirement for Arbitrary Logic

### PR DESCRIPTION
**Pull request by wyc**
These changes allow Submission Requirement to support arbitrary logic to
support the use case described in the comment here:

https://github.com/decentralized-identity/presentation-exchange/pull/18/files#diff-9308a2575ae14de31b25f4459ecfc604R57

This is what it would look like:

```json
{"submission_requirement": {
  "name": "((any two from group A) AND (any one from group B)) OR (any three from group B)",
  "purpose": "Required submission",
  "rule": "pick",
  "count": 1,
  "from": [
    {
      "name": "(any two from group A) AND (any one from group B)",
      "rule": "all",
      "from": [
	{
	  "name": "any two from group A",
	  "rule": "pick",
	  "count": 2,
	  "from": "A"
	},
        {
	  "name": "any one from group B",
	  "rule": "pick",
	  "count": 1,
	  "from": "B"
        }
      ]
    },
    {
      "name": "any three from group B",
      "rule": "pick",
      "count": 3,
      "from": "B"
    }
  ]
}}
```

Note that while `pick` can require "exactly three" we do not currently
support a rule that allows for "three or more", but this could be added
in the future.

Added a test to validate that an instance in the new format satisfying
the use case above also adheres to the JSON Schema:

```json
{                                                           
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {                                               
    "submission_requirement": {
      "type": "object",              
      "required": ["rule", "from"],
      "properties": {                   
        "name": { "type": "string" },
        "purpose": { "type": "string" },
        "rule": { "type": "string" },
        "count": { "type": "integer", "minimum": 1 },
        "from": {
          "anyOf": [
            { "type": "string" },                            
            {
	      "type": "array",
	      "minItems": 1,
              "items": { "$ref": "#/definitions/submission_requirement" }
            }
          ]
        }
      }
    }
  },
  "type": "object",
  "properties": {
    "submission_requirement": { "$ref": "#/definitions/submission_requirement" }
  }
}
```

### Issues to talk about on the next working session:
* Should we explore MongoDB-style querying again? What would `pick` look like? This is probably bad for many reasons:
```
{"$and": [
  {"$pick-3": "A"},
  {"$all: "B"},
]}
```
* Should we have different operators and/or fields depending on if we are descending into the tree or terminating on a leaf group? Or is overloading okay?
* Is it okay we got rid of the array of groups?
* Is it okay we got rid of the rule object?

### CODE REVIEW

File: spec/spec.md [hunk 1]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+    name: "Credential issuance requirements",
+    purpose: "Verify banking, employment, and citizenship information.",
+    rule: "all",
+    from: [
+      {
+        "name": "Banking Information",
+        "purpose": "We need to know if you have an established banking history.",
+        "rule": "pick",
+        "count": 1,
+        "from": ["A"]
```
Comment (brentzundel) [inline spec/spec.md:269]:
```suggestion
        "from": "A"
```
Should this match the example above?

File: spec/spec.md [hunk 2]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+      },
+      {
+        "name": "Employment Information",
+        "purpose": "We need to know that you are currently employed.",
+        "rule": "all",
+        "from": ["B"]
```
Comment (brentzundel) [inline spec/spec.md:275]:
```suggestion
        "from": "B"
```

File: spec/spec.md [hunk 3]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+      }
+      {
+        "name": "Citizenship Information",
+        "rule": "pick",
+        "count": 1,
+        "from": ["C"]
```
Comment (brentzundel) [inline spec/spec.md:281]:
```suggestion
        "from": "C"
```

File: spec/spec.md [hunk 4]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
 
-#### Requirement Rules
+	(a) A `group` string matching one or more of the _Input Descriptor_
+	objects in the `input_descriptors`
 
-[_Requirement Rules_](#requirement-rules){id="requirement-rules"} are used within _Requirement Objects_ to describe the specific combinatorial rule that must be applied to submit a particular subset of required inputs. An implementation ****MUST**** support the following standard types:
+	(b) An array of nested _Submission Requirement_ objects.
+
+3. The object ****MAY**** contain a `name` property, and if present, its value
+   ****MUST**** be a string of the creator's choosing, which ****MAY**** be
```
Comment (brentzundel) [inline spec/spec.md:307]:
```suggestion
   ****MUST**** be a string which ****MAY**** be
```

File: spec/spec.md [hunk 5]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+   used by a consuming User Agent to display the general name of the
+   requirement set to a user.
+4. The object ****MAY**** contain a `purpose` property, and if present, its
```
Comment (brentzundel) [inline spec/spec.md:310]:
```suggestion
4. The object ****MAY**** contain a `purpose` property and, if present, its
```

File: spec/spec.md [hunk 6]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+    "name": "Either all from group A or two from group B",
+    "purpose": "Either way works",
+    "rule": "pick",
     "count": 1,
-    "from": ["B", "C"]
+    "from": [
+      {"rule": "all",  "from": "A"},
+      {"rule": "pick", "from": "B", "count": 2},
+    ]
+  }
```
Comment (brentzundel) [inline spec/spec.md:411]: I like this use of nesting which allows a `pick` rule to operate as a logical `or`

File: spec/spec.md [hunk 7]
```diff
@@ -231,89 +231,258 @@ Presentation Definitions are objects generate to articulate what proofs an entit
...
+      - If `rule is `"all"`, then this _Submission Requirement_ is satisfied if
+	and only if all nested _Submission Requirement_ objects are satisfied.
+      - If `rule` is `"pick"`, then this _Submission Requirement_ is satisfied
+	if and only if the number of satisfied nested _Submission Requirement_
+	objects is exactly `count`.
 
 ### Input Descriptors
 
-_Input Descriptors_ are objects used to describe the proofing inputs a Verifier requires of a Subject before they will proceed with an interaction. _Input Descriptor_ objects contain a schema URI that links to the schema if the required input data, constrains on data values, and an explanation of why a certain set or item of data is being requested:
+_Input Descriptors_ are objects used to describe the proofing inputs a Verifier requires of a Subject before they will proceed with an interaction. _Input Descriptor Objects_ contain a schema URI that links to the schema if the required input data, constrains on data values, and an explanation of why a certain set or item of data is being requested:
```
Comment (brentzundel) [inline spec/spec.md:485]:
```suggestion
_Input Descriptors_ are objects used to describe the proofing inputs a Verifier requires
of a Subject before they will proceed with an interaction. _Input Descriptor Objects_
contain a schema URI that links to the schema of the required input data, constraints
on data values, and an explanation why a certain item or set of data is being requested:
```

File: spec/spec.md [hunk 8]
```diff
@@ -365,7 +534,7 @@ _Input Descriptors_ are objects used to describe the proofing inputs a Verifier
 
 #### Input Descriptor Objects
 
-_Input Descriptors_ are objects that describe what type of input data/credential, or sub-fields thereof, is required for submission to the Verifier. _Input Descriptor_ objects are composed as follows:
+_Input Descriptors_ are objects that describe what type of input data/credential, or sub-fields thereof, is required for submission to the Verifier. _Input Descriptor Objects_ are composed as follows:
 
   - The object ****MUST**** contain an `id` property, and if present, its value ****MUST**** be a unique identifying string that does not conflict with the `id` of another _Input Descriptor_ in the same _Presentation Definition_ object.
```
Comment (brentzundel) [inline spec/spec.md:539]:
```suggestion
  - The object ****MUST**** contain an `id` property. The value of the `id` property ****MUST**** be a unique identifying string that does not conflict with the `id` of another _Input Descriptor_ in the same _Presentation Definition_ object.
```

Comment (brentzundel) [reply spec/spec.md:539]: I know this isn't a comment on a change you've recommended, but it needs fixing.

File: spec/spec.md (hunk 1 above)
Comment (wyc) [reply spec/spec.md:269]: yes!

File: spec/spec.md (hunk 6 above)
Comment (wyc) [reply spec/spec.md]: image

**Comment by wyc** [comment]
Note to prospective merger: please squash and rebase-merge with the main commit message so we can keep a nice linear history with a single additional commit. Thanks.

**Comment by csuwildcat** [comment]
I think this does what we want - any other comments or changes from others before you're ready to merge? (cc: @OR13, @glcohen)

**Comment by wyc** [comment]
Thank you @brentzundel and @OR13 for the review + comments!

**Pull request merged**